### ISSUE TITLE
Task-review overview

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#702]
Me and @mfranzon tried to come up with an improved organization of tasks. Our understanding is that we are not ready to define strict access control, because we have no visibility into many-{users/teams} Fractal usage. We then propose to proceed in a way which will allow us to incrementally move towards strict access control when it will be better defined.

# TL;DR of our proposal
1. We add a `task.owner` attribute (or an attribute with a similar name), which is a standard string attribute (and not an actual DB relationship to the `user` table). This can be `None`, representing a "common" task.
2. Tasks coming from automatic collection are marked as "common", while tasks coming from the task-add endpoint are marked with some non-null `owner` value.
3. The `source` attribute of tasks coming from automatic collection is better defined, in a way which includes more metadata and then allows mutliple client-side features (e.g. https://github.com/fractal-analytics-platform/fractal/issues/345 and https://github.com/fractal-analytics-platform/fractal-web/issues/149).

# A decision to be taken

In the following, we imagine that the `task.owner` attribute is fully free, and can be specified by hand by the user adds a task - with no checks. This gives us a lot of freedom for how to evolve, but it requires the user to always specify it.

The rationale here is that users who add custom tasks are some kind of "power users", and we can ask them to accept the additional burden of always adding their nickname in a form.

If this seems too cumbersome, then there are alternative ways to proceed:
1) Automatically set it to the email address of the user (which is already known). We don't like this option, cause it exposes a private information which would be visible to all users.
2) Automatically set it to the `slurm_user` attribute of the user. We don't like this option, because it's bound to an attribute which is optional (it only becomes necessary if you actually submit a job, but a user could spend time on Fractal without even having a SLURM account.. or we could have a `local` backend and no `slurm_user` attributes).
3) We create an additional field of the `user` DB table, which is a Fractal username, and then we use it for the purpose described above (or possibly for other purposes in the future). This would be an attribute which a user has no problem with sharing (clearly different from sharing an actual email address).

# Relevant use cases

(U1) A user creates a Fractal task that is meant to be available to all users, and this task is clearly identified as "common" in Fractal (in the database and in the web client).
(U2) The same as U1 can be achieved for a Fractal-compatible Python package of tasks; in this case, Fractal takes care of: preparing a dedicated venv, installing the package, creating the correct task metadata, and adding the tasks to Fractal. Context: the Fractal admin install fractal-tasks-core or a similar task package.
(U3) A user creates a Fractal task that is not meant to be "common", and this information is clearly visible (in the database and in the web client). Context: the user wants to develop a new task.
(U4) The user proceeds as in U3, and then they can easily make the task a "common" one. Context: the development of a new task reached a stable state.
(U5) The user proceeds as in U3, and then they can mark the task as available to a specific subgroup of users. Context: a large Fractal instance shared by multiple teams.

For each use case, there are choices to be made about permissions, e.g.:
* Which user can perform U1 or U2?
* After U3, which users have access to that task?
* Which users can perform U4?
* How is access control set up for U5?

# Re: access control

Strict access control can be set up in the same way as we do for projects/users (that is, a many-to-many user/task relationship in the DB, and additional access-control checks in all relevant endpoints), or in more explicit ways (DB attributes which are not actual relationships, like a `task.owner` column, and checks in the relevant endpoints). Both approaches are doable, and they both come with a pros/cons:
* The user/task DB relationship is cumbersome to implement and verify, especially when it's mixed with the project/user relationship. On the other hand, it leaves no room to ambiguity: a user can only interact with a task if they are in a DB relationship.
* The loose-owner-attribute approach requires human knowledge of a set of access control rules, to be then included in the relevant endpoints. On the other hand, implementation is simple, and (most importantly) this approach lends itself to an incremental implementation (e.g. it can be used to only set _some_ access-control rules, rather than all of them at once).

Given that we don't have a good visibility on the longer-term access-control needs, we propose to follow the loose-owner-attribute approach. This gets us started right away, with no access control at all, and will allows us to introduce some endpoint checks as soon as we take decisions. The "burden" of actual access control based on DB relationship should first be justified by additional insights.

Note: the current "private" flag of the task-collection endpoint is useless (it just changes the folder name where tasks are installed) and should be removed, until we define what a private task is (ref #223).

# Source attribute for automatically-collected tasks

We came up with an enriched version of the `task.source` attribute, for tasks coming from automatic collection from a python package.
The source will be a string with the following fields:
* `collection_type` (required), e.g. `pip_remote` or `pip_local` (or maybe `conda` or others)
* `package_name` (required), e.g. `fractal_tasks_core`. The ground truth value is the one in the output of `pip show`, rather than the user input.
* `package_version` (required), e.g. `0.1.2`. The ground truth value is the one in the output of `pip show`, rather than the user input (which is actually optional).
* `python_version` (required), e.g. `py3.8`. This only takes into account `major.minor`. For the moment it is not fully supported (see #659), but it will be necessary to have it in the source.
* `package_extras` (optional), e.g. `some_extra1,some_extra2`. Note an important difference from package name/version fields: this field cannot be extracted easily from `pip show` (we could rely on `pip inspect`, but it's a somewhat recent feature that I'd rather not include as crucial). Therefore we'll use the user input: if the user asks for `package_extras="something_invalid"`, then we'll write `"something_invalid"` in the collected-tasks source.
* `task_identifier`. This needs to be either the slug of the task name, or the executable name.

An example of the chosen source specs is
```
pip_remote:fractal-tasks-core:0.9.3:py3.9:extra1,extra2,extra3:create_ome_zarr
```
This source should be enough to enable some required client-side features (e.g. https://github.com/fractal-analytics-platform/fractal/issues/345 and https://github.com/fractal-analytics-platform/fractal-web/issues/149). In fact it will have more information than needed (e.g. the Python version and the package extras), but those are needed to make it as uniquely defined as possible.

### Re: portability

As compared to previous versions, we are making the source syntax more rich. Using `fractal-tasks-core` as an example, this means that the `cellpose_segmentation` task of 0.9.4 on a given Fractal instance won't automatically correspond to the one on another instance, if some fields (e.g. the Python version) differ. This introduces some friction when importing a workflow into a different Fractal instance, because if the two instance admins installed `fractal-tasks-core` with different Python version there will need to be some manual editing of the to-be-imported workflow JSON file (updating all Python versions in the task sources).

This cost cannot be easily avoided, if we want to be able to control the Python version of tasks. And this is something we most likely will have to do, given that (e.g.) cellpose still comes with an official "We recommend python 3.8" advice.

### Re: `pip_remote` vs `pip_local`

First, a maybe-not-so-obvious clarification: installing a task package from a local wheel file (the `pip_local` collection type) is not a local operation. If the task has even just one dependency, then that one will be fetched from PyPI (or other sources if specified). This may suggest that `pip_remote` and `pip_local` collection types are actually the same thing, and we could remove this information from the source, but it's misleading.

There are two important specific cases where `pip_local` is special:
1) The very relevant use case where a package is private. In this case, all dependencies may be available on PyPI, but the actual package will not (e.g. it is developed locally, or on a private remote repository).
2) The package is a local fork of an existing PyPI package, e.g. a local fork of `fractal-tasks-core` that fixes a small bug which is only relevant in a given lab and which for the moment is not available on PyPI. Then the package will have the same name/version of the PyPI one, but we cannot risk having the same `source` attribute, because if it is local then Fractal cannot guarantee what is in there.

Note that case 2 is not expected to be frequent, but case 1 probably will be relevant. This is enough to justify the presence of `pip_local/pip_remote` in the source.

### Re: why a string and not many db columns?

Question: Why don't we split the source string into several attributes?
Answer: That would be too strict, for instance if we want to add a single "common" task that doesn't correspond to a Python package (e.g. a bash task that calls the `bioconvert` CLI). By keeping everything in the `source` attribute, we keep the freedom of doing so, without breaking anything on the fractal-server side. The discussion on how to handle it will then happen client-side.

In other words: fractal-server should not rely on the actual `source` structure, but it should only write it correctly for automatically-collected Python tasks. The different clients will then consume these sources to offer additional features.

### Re: uniqueness constraints

In the task table, we should make sure that the `(owner, source)` pair is unique. This means that two users can both have a `source=test` task, but there cannot be two `pip_remote:fractal-tasks-core:0.9.3:py3.9:extra1,extra2,extra3:create_ome_zarr` common tasks.

# Preliminary to do list (some should be converted in issues):

EDIT: updated and moved to another comment  below

**Comment by tcompa** [comment]
To review in view of https://github.com/fractal-analytics-platform/fractal/issues/345#issuecomment-1559595699:

* Is the `version` information important enough that we want to include it for all tasks (as an independent DB column), rather than only as part of the `source` attribute of automatically-collected python-package tasks?

**Comment by tcompa** [comment]
#704 closed many of the issues, but a few are still left in the to-do list

**Comment by tcompa** [comment]
> Define what behaviors depend on the task being standard or not (e.g. import/export of a workflow is not guaranteed to work for non-standard tasks - we should at least raise a warning, or maybe even fail).
> EDIT: depending on the choice made in https://github.com/fractal-analytics-platform/fractal-server/issues/721, we may not have to raise any warning/error.

As of 01f7486, we have a warning when exporting a workflow which includes workflow tasks. This will be enough for the current issue, and I'm opening a new issue to decide whether this should be simply a server-side warning or it should actually affect the server response during workflow import/export.
